File: desktop/src/lib/agent-browsers.ts
/* ------------------------------------------------------------------ */
/*  Types                                                              */
/* ------------------------------------------------------------------ */

export interface BrowserProfile {
  id: string;
  agent_name: string | null;
  profile_name: string;
  node: string;
  status: "stopped" | "running" | "error";
  container_id: string | null;
  created_at: number;
  updated_at: number;
}

export interface LoginStatus {
  x: boolean;
  github: boolean;
  youtube: boolean;
  reddit: boolean;
}

export interface CookieEntry {
  name: string;
  value: string;
  domain: string;
  path: string;
  expires: number;
  httpOnly: boolean;
  secure: boolean;
}

/* ------------------------------------------------------------------ */
/*  Helpers                                                            */
/* ------------------------------------------------------------------ */

async function fetchJson<T>(url: string, fallback: T, init?: RequestInit): Promise<T> {
  try {
    const res = await fetch(url, { ...init, headers: { Accept: "application/json", ...init?.headers } });
    if (!res.ok) return fallback;
    const ct = res.headers.get("content-type") ?? "";
    if (!ct.includes("application/json")) return fallback;
    return await res.json();
  } catch {
    return fallback;
  }
}

async function postJson<T>(url: string, body: unknown, fallback: T): Promise<T> {
  return fetchJson(url, fallback, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
}

async function putJson<T>(url: string, body: unknown, fallback: T): Promise<T> {
  return fetchJson(url, fallback, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
}

/* ------------------------------------------------------------------ */
/*  Profiles                                                           */
/* ------------------------------------------------------------------ */

export async function listProfiles(agentName?: string): Promise<BrowserProfile[]> {
  const qs = agentName ? `?agent_name=${encodeURIComponent(agentName)}` : "";
  const data = await fetchJson<{ profiles: BrowserProfile[] }>(
    `/api/agent-browsers/profiles${qs}`,
    { profiles: [] },
  );
  return Array.isArray(data.profiles) ? data.profiles : [];
}

export async function createProfile(
  name: string,
  agentName?: string,
  node?: string,
): Promise<{ id: string; status: string } | null> {
  try {
    const res = await fetch("/api/agent-browsers/profiles", {
      method: "POST",
      headers: { "Content-Type": "application/json", Accept: "application/json" },
      body: JSON.stringify({ profile_name: name, agent_name: agentName ?? null, node: node ?? "local" }),
    });
    if (!res.ok) return null;
    const ct = res.headers.get("content-type") ?? "";
    if (!ct.includes("application/json")) return null;
    return await res.json();
  } catch {
    return null;
  }
}

export async function deleteProfile(id: string): Promise<boolean> {
  try {
    const res = await fetch(`/api/agent-browsers/profiles/${encodeURIComponent(id)}`, {
      method: "DELETE",
      headers: { Accept: "application/json" },
    });
    return res.ok;
  } catch {
    return false;
  }
}

export async function deleteProfileData(id: string): Promise<boolean> {
  try {
    const res = await fetch(`/api/agent-browsers/profiles/${encodeURIComponent(id)}/data`, {
      method: "DELETE",
      headers: { Accept: "application/json" },
    });
    return res.ok;
  } catch {
    return false;
  }
}

/* ------------------------------------------------------------------ */
/*  Browser lifecycle                                                  */
/* ------------------------------------------------------------------ */

export async function startBrowser(id: string): Promise<BrowserProfile | null> {
  const data = await postJson<BrowserProfile | null>(
    `/api/agent-browsers/profiles/${encodeURIComponent(id)}/start`,
    {},
    null,
  );
  return data;
}

export async function stopBrowser(id: string): Promise<BrowserProfile | null> {
  const data = await postJson<BrowserProfile | null>(
    `/api/agent-browsers/profiles/${encodeURIComponent(id)}/stop`,
    {},
    null,
  );
  return data;
}

/* ------------------------------------------------------------------ */
/*  Screenshot                                                         */
/* ------------------------------------------------------------------ */

export async function getScreenshot(id: string): Promise<string | null> {
  const data = await fetchJson<{ data?: string }>(
    `/api/agent-browsers/profiles/${encodeURIComponent(id)}/screenshot`,
    {},
  );
  return data.data ?? null;
}

/* ------------------------------------------------------------------ */
/*  Cookies                                                            */
/* ------------------------------------------------------------------ */

export async function getCookies(
  agent: string,
  profile: string,
  domain: string,
): Promise<CookieEntry[]> {
  const qs = `?domain=${encodeURIComponent(domain)}`;
  const data = await fetchJson<{ cookies: CookieEntry[] }>(
    `/api/agent-browsers/${encodeURIComponent(agent)}/${encodeURIComponent(profile)}/cookies${qs}`,
    { cookies: [] },
  );
  return Array.isArray(data.cookies) ? data.cookies : [];
}

/* ------------------------------------------------------------------ */
/*  Login status                                                       */
/* ------------------------------------------------------------------ */

export async function getLoginStatus(id: string): Promise<LoginStatus | null> {
  try {
    const res = await fetch(
      `/api/agent-browsers/profiles/${encodeURIComponent(id)}/login-status`,
      { headers: { Accept: "application/json" } },
    );
    if (!res.ok) return null;
    const ct = res.headers.get("content-type") ?? "";
    if (!ct.includes("application/json")) return null;
    return await res.json();
  } catch {
    return null;
  }
}

/* ------------------------------------------------------------------ */
/*  Agent assignment / node migration                                  */
/* ------------------------------------------------------------------ */

export async function assignAgent(id: string, agentName: string): Promise<BrowserProfile | null> {
  const data = await putJson<BrowserProfile | null>(
    `/api/agent-browsers/profiles/${encodeURIComponent(id)}/assign`,
    { agent_name: agentName },
    null,
  );
  return data;
}

export async function moveToNode(id: string, node: string): Promise<BrowserProfile | null> {
  const data = await putJson<BrowserProfile | null>(
    `/api/agent-browsers/profiles/${encodeURIComponent(id)}/move`,
    { node },
    null,
  );
  return data;
}

File: tinyagentos/routes/agent_browsers.py
from __future__ import annotations

"""API routes for Agent Browsers -- persistent Chromium profiles per agent.

All routes live under /api/agent-browsers/.  The router reads
``request.app.state.agent_browsers`` which must be an
``AgentBrowsersManager`` instance.
"""

import logging

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse, Response
from pydantic import BaseModel

logger = logging.getLogger(__name__)

router = APIRouter()


# ------------------------------------------------------------------
# Pydantic models
# ------------------------------------------------------------------

class CreateProfileRequest(BaseModel):
    profile_name: str
    agent_name: str | None = None
    node: str = "local"


class AssignRequest(BaseModel):
    agent_name: str


class MoveRequest(BaseModel):
    node: str


# ------------------------------------------------------------------
# Profiles -- CRUD
# ------------------------------------------------------------------

@router.get("/api/agent-browsers/profiles")
async def list_profiles(request: Request, agent_name: str | None = None):
    """List all browser profiles, optionally filtered by agent_name."""
    mgr = request.app.state.agent_browsers
    profiles = await mgr.list_profiles(agent_name=agent_name)
    return {"profiles": profiles, "count": len(profiles)}


@router.post("/api/agent-browsers/profiles")
async def create_profile(request: Request, body: CreateProfileRequest):
    """Create a new browser profile."""
    mgr = request.app.state.agent_browsers
    try:
        profile = await mgr.create_profile(
            profile_name=body.profile_name,
            agent_name=body.agent_name,
            node=body.node,
        )
        return profile
    except Exception as exc:
        logger.exception("create_profile failed: %s", exc)
        return JSONResponse({"error": str(exc)}, status_code=500)


@router.get("/api/agent-browsers/profiles/{profile_id}")
async def get_profile(request: Request, profile_id: str):
    """Fetch a single profile by id."""
    mgr = request.app.state.agent_browsers
    profile = await mgr.get_profile(profile_id)
    if profile is None:
        return JSONResponse({"error": "not found"}, status_code=404)
    return profile


@router.delete("/api/agent-browsers/profiles/{profile_id}")
async def delete_profile(request: Request, profile_id: str):
    """Delete a profile (stops container if running, does NOT delete volume)."""
    mgr = request.app.state.agent_browsers
    deleted = await mgr.delete_profile(profile_id)
    if not deleted:
        return JSONResponse({"error": "not found"}, status_code=404)
    return {"status": "deleted", "id": profile_id}


@router.delete("/api/agent-browsers/profiles/{profile_id}/data")
async def delete_profile_data(request: Request, profile_id: str):
    """Delete the Docker volume backing a profile (irreversible)."""
    mgr = request.app.state.agent_browsers
    await mgr.delete_profile_data(profile_id)
    return {"status": "volume_deleted", "id": profile_id}


# ------------------------------------------------------------------
# Browser lifecycle
# ------------------------------------------------------------------

@router.post("/api/agent-browsers/profiles/{profile_id}/start")
async def start_browser(request: Request, profile_id: str):
    """Start the browser for a profile."""
    mgr = request.app.state.agent_browsers
    ok = await mgr.start_browser(profile_id)
    if not ok:
        return JSONResponse({"error": "failed to start or profile not found"}, status_code=400)
    return {"status": "started", "id": profile_id}


@router.post("/api/agent-browsers/profiles/{profile_id}/stop")
async def stop_browser(request: Request, profile_id: str):
    """Stop the browser for a profile."""
    mgr = request.app.state.agent_browsers
    ok = await mgr.stop_browser(profile_id)
    if not ok:
        return JSONResponse({"error": "profile not found"}, status_code=404)
    return {"status": "stopped", "id": profile_id}


# ------------------------------------------------------------------
# Browser inspection
# ------------------------------------------------------------------

@router.get("/api/agent-browsers/profiles/{profile_id}/screenshot")
async def get_screenshot(request: Request, profile_id: str):
    """Return a PNG screenshot of the browser (CDP, 30s cache)."""
    mgr = request.app.state.agent_browsers
    png = await mgr.get_screenshot(profile_id)
    if png is None:
        return JSONResponse({"error": "screenshot unavailable"}, status_code=404)
    return Response(content=png, media_type="image/png")


@router.get("/api/agent-browsers/{agent_name}/{profile_id}/cookies")
async def get_cookies(request: Request, agent_name: str, profile_id: str, domain: str = ""):
    """Return cookies for a domain from the Chromium cookie store."""
    mgr = request.app.state.agent_browsers
    cookies = await mgr.get_cookies(profile_id, domain)
    return {"cookies": cookies, "domain": domain}


@router.get("/api/agent-browsers/profiles/{profile_id}/login-status")
async def get_login_status(request: Request, profile_id: str):
    """Return per-site login status (x, github, youtube, reddit)."""
    mgr = request.app.state.agent_browsers
    status = await mgr.get_login_status(profile_id)
    return status


# ------------------------------------------------------------------
# Assignment & routing
# ------------------------------------------------------------------

@router.put("/api/agent-browsers/profiles/{profile_id}/assign")
async def assign_agent(request: Request, profile_id: str, body: AssignRequest):
    """Assign a profile to an agent."""
    mgr = request.app.state.agent_browsers
    ok = await mgr.assign_agent(profile_id, body.agent_name)
    if not ok:
        return JSONResponse({"error": "profile not found"}, status_code=404)
    return {"status": "assigned", "id": profile_id, "agent_name": body.agent_name}


@router.put("/api/agent-browsers/profiles/{profile_id}/move")
async def move_to_node(request: Request, profile_id: str, body: MoveRequest):
    """Move a profile to a different cluster node."""
    mgr = request.app.state.agent_browsers
    ok = await mgr.move_to_node(profile_id, body.node)
    if not ok:
        return JSONResponse({"error": "profile not found"}, status_code=404)
    return {"status": "moved", "id": profile_id, "node": body.node}
